crates/rpc/rpc/src/eth/filter.rs

```diff
@@ -1014,48 +1023,82 @@ struct RangeBlockMode<
 }
 
 impl<
         Eth: RpcNodeCoreExt<Provider: BlockIdReader, Pool: TransactionPool> + EthApiTypes + 'static,
     > RangeBlockMode<Eth>
 {
     async fn next(&mut self) -> Result<Option<ReceiptBlockResult<Eth::Provider>>, EthFilterError> {
-        if let Some(result) = self.next.pop_front() {
-            return Ok(Some(result));
-        }
+        loop {
+            // First, try to return any already processed result from buffer
+            if let Some(result) = self.next.pop_front() {
+                return Ok(Some(result));
+            }
 
-        let Some(next_header) = self.iter.next() else {
-            return Ok(None);
-        };
+            // Try to get a completed task result if there are pending tasks
+            if let Some(task_result) = self.pending_tasks.next().await {
+                let chunk_results = task_result?;
+                if !chunk_results.is_empty() {
+                    // Add all results to buffer, first one will be returned in next loop iteration
+                    self.next.extend(chunk_results);
+                    // Continue loop to return the first result from buffer
+                    continue;
+                }
+                // Task completed with empty results, continue to check for more work
+                continue;
+            }
 
-        let mut range_headers = Vec::with_capacity(self.max_range);
-        range_headers.push(next_header);
+            // No pending tasks - try to generate new work
+            let Some(next_header) = self.iter.next() else {
+                // No more headers to process
+                return Ok(None);
+            };
 
-        // Collect consecutive blocks up to max_range size
-        while range_headers.len() < self.max_range {
-            let Some(peeked) = self.iter.peek() else { break };
-            let Some(last_header) = range_headers.last() else { break };
+            let mut range_headers = Vec::with_capacity(self.max_range);
+            range_headers.push(next_header);
 
-            let expected_next = last_header.header().number() + 1;
-            if peeked.header().number() != expected_next {
-                break; // Non-consecutive block, stop here
-            }
+            // Collect consecutive blocks up to max_range size
+            while range_headers.len() < self.max_range {
+                let Some(peeked) = self.iter.peek() else { break };
+                let Some(last_header) = range_headers.last() else { break };
 
-            let Some(next_header) = self.iter.next() else { break };
-            range_headers.push(next_header);
-        }
+                let expected_next = last_header.number() + 1;
+                if peeked.number() != expected_next {
+                    debug!(
+                        target: "rpc::eth::filter",
+                        last_block = last_header.number(),
+                        next_block = peeked.number(),
+                        expected = expected_next,
```

Comment:
would be good to include the range size here
```suggestion
                        expected = expected_next,
                        range_size = range_headers.len(),
```